### ISSUE TITLE
Including the vendor and app bundles in Liferay

### ISSUE THREAD

**Issue by taskiesa-visma** [lundegaard/react-union#41]
Do you have examples or advice on how to include the vendor and app bundles (and ensuring that an app's bundle is only loaded once, when there are multiple instances of a portlet on a page) in Liferay? Currently I'm simply adding script tags to a portlet embedded in my theme, but I was wondering if you're using a different method.

**Comment by jamescoq** [comment]
You have basically 2 options to do that. 
1) include all your js bundles into Liferay theme 
2) use AMD module loader to include the vendor and app bundles, https://dev.liferay.com/develop/tutorials/-/knowledge_base/7-0/liferay-amd-module-loader#defining-a-module. Once you have all JS loaded in Liferay, then require the entry JS module in your portlet.

The option 2 seems to be better aligned with Liferay. We have some examples, but we need to consult with our company whether we can provide them.

**Comment by taskiesa-visma** [comment]
Thanks for the input! I actually also considered the first option, which would require building a portal_normal file from a template and deploying the theme when bundles are updated.

I'll look into the second option as well.

**Comment by tommmyy** [comment]
In addition to 1) 
I use dynamically generated ftl template (generated by html-webpack-plugin), that is included in current theme. 

`union.ftl` (generated by html-webpack-plugin):
```ftl
<script type="text/javascript" src="/o/liferay-7-theme/js/react-apps/manifest.bundle.js"></script><script type="text/javascript" src="/o/liferay-7-theme/js/react-apps/vendor.bundle.js"></script><script type="text/javascript" src="/o/liferay-7-theme/js/react-apps/entry.bundle.js"></script>
```
and then include in `portal_normal.ftl`:

```ftl
<head>
	<#include "${full_templates_path}/union.ftl" />
</head>

```

If you use Liferay AMD loader, it is essential to have your JS module in one osgi module.

I can provide more info if you like. Maybe to add little recipe to README.md (@jamescoq)

**Comment by jamescoq** [comment]
@taskiesa-visma your assumptions are correct. 

As for the second option. In the AMD module loader osgi module you have to specify which modules to load. It is enough to declare vendor modul and entry module (with dependency to vendor module). You do not need to specify individual application chunks in the config.js, but make sure to include them in the amd module loader :) ... the chunk files and other static resource that you will include in the AMD module loader will be then accessible on path: localhost:8080/o/[your-amd-module-loader-name]/js/ . Therefore you have to specify correct publicPath in union.config.js
```
apps: [
		{
			name: 'SampleApp',
			path: path.resolve(__dirname, './src/apps/SampleApp'),
			proxy: {
				publicPath: '/o/[your-amd-module-loader-name]/js/',
				target: 'http://localhost:8080',
			},
		},
	],
```

**Comment by taskiesa-visma** [comment]
> I can provide more info if you like. Maybe to add little recipe to README.md

That'd be great, since Liferay examples are already on your roadmap.

I ran into some JS errors when including an app bundle. The first one solved itself after I updated my portal from 7.0.4-ga5 to 7.1.0-m1. Have you encountered these?

```
{Liferay JS modules path}&/o/frontend-js-web/aui/querystring-parse/querystring-parse-min.js:7 Uncaught TypeError: Cannot read property 'all' of null

KTBaseApp.4a921284.bundle.js:1 Uncaught Error: only one instance of babel-polyfill is allowed
    at Object.<anonymous> (KTBaseApp.4a921284.bundle.js:1)
    at Object.2586 (KTBaseApp.4a921284.bundle.js:1)
    at a (vendor.7ba2697d.bundle.js:1)
    at Object.2585 (KTBaseApp.4a921284.bundle.js:1)
    at a (vendor.7ba2697d.bundle.js:1)
    at window.webpackJsonp (vendor.7ba2697d.bundle.js:1)
    at KTBaseApp.4a921284.bundle.js:1
```

The second one doesn't break anything, but I'd like to avoid it as well. It has to do with babel-polyfill being included by both Liferay and the app bundle?

**Comment by tommmyy** [comment]
The babel-polyfill is included twice 
- by react-union-scripts (currently there is no option to exclude it) 
- and in Liferay (https://dev.liferay.com/develop/tutorials/-/knowledge_base/7-0/preparing-your-javascript-files-for-es2015)

**Comment by tommmyy** [comment]
@taskiesa-visma  I addressed that issue in #42

**Comment by taskiesa-visma** [comment]
Thanks again for the quick replies @tommmyy and @jamescoq. I'll close the issue and look forward to new releases.
